### PR TITLE
docs(notebook): add ERA5 CONUS fire-weather demonstration

### LINKED ISSUE #811: FIRE-19 — ERA5 CONUS fire-weather demonstration notebook
**Labels:** `documentation`, `fire-weather`, `demo`
**Milestone:** `v2.8 — Fire weather diagnostics`
**Depends on:** #807, #801

### Summary

Add a notebook under `notebooks/` computing the full fire family over CONUS or
North America from ERA5, using xarray and Dask, and producing seasonal and
extreme-percentile fire-weather climatologies. NASA GISS's Global Fire WEather
Database, which is built on the FWI System, provides a useful conceptual
comparison target.

### Tasks

- [ ] Data access section using a public ERA5 source, with a small pre-staged
      subset committed so the notebook runs without credentials
- [ ] Compute KBDI, all CFFWIS components, Fosberg and HDW over the domain
- [ ] Produce seasonal climatology and 90th/95th percentile fire-weather maps
- [ ] Show the drought-to-fire-weather chain by overlaying SPI-3, EDDI and DC
      for a documented fire season
- [ ] Include a qualitative comparison against GFWED patterns with an explicit
      statement that it is qualitative
- [ ] Add timing and memory notes demonstrating the Dask scaling story

### Acceptance criteria

- Notebook executes end to end on the committed subset in under 10 minutes
- Included in docs build or linked from the wildfire applications page
- No unqualified claims of agreement with GFWED

### References

- NASA GISS Global Fire WEather Database — https://data.giss.nasa.gov/impacts/gfwed/

### PR DESCRIPTION
**Pull request by monocongo**
## Summary

- Add `scripts/prepare_fire_demo_inputs.py`, which reads the anonymously accessible ARCO-ERA5 Zarr store (no credentials), subsets a 1.5 degree CONUS box to daily surface fields over 2018-2020 and daily pressure-level profiles over the 2020 fire season, and caches per-variable downloads under the git-ignored `data/fire-demo/`.
- Add `notebooks/fire_weather_demo.ipynb`: KBDI over three years, the seven CFFWIS components, Fosberg FFWI, and HDW over the 2020 season, with seasonal means, 90th/95th percentile maps, the 7 September 2020 HDW, and a drought-to-fire-weather panel. Executes end to end in about three seconds on the prepared subset, offline.
- Link the notebook from `docs/wildfire_applications.md` and restate there that it is a worked example, not a reference product.
- Add `tests/test_prepare_fire_demo_inputs.py` covering the daily aggregation (including the year-boundary precipitation bin), the derived relative humidity, the above-ground level selection, the level-file derivation, the cache, publishing, the manifest, and the coordinate conversion.

## Scope decisions

The issue assumed a small committed subset. The repository instead keeps reproducible notebook inputs out of git (`data/e2e/` precedent, `scripts/prepare_e2e_inputs.py`), so the data is downloaded and cached locally. Two further reductions, each stated in the notebook's "What this demo does not claim" section:

- **SPI and EDDI overlays are deferred.** The xarray adapters require at least 30 years of non-NaN calibration data; the subset carries three. The drought-to-fire chain is shown with DC and KBDI instead, and the 30-year-baseline overlay is tracked by #980.
- **Coarse grid, one season, daily summaries.** 1.5 degree cells, the 2020 fire season for the level-based indices, daily means/maxima/sums timestamped 12:00 in place of noon local-standard-time observations, and relative humidity derived from specific humidity at the lowest level above the surface (ERA5 publishes no 2 m dewpoint in this store). KBDI's mean annual precipitation comes from the three prepared years because the adapter refuses to infer one from fewer than 30. The CFFWIS moisture codes are cold-started from the literature seeds, so DC reports one season's drying.

## Validation

- `uv run ruff check src/ tests/` (plus `scripts/prepare_fire_demo_inputs.py`)
- `uv run ruff format --check src/ tests/` (plus the new script)
- `uv run mypy src/` -- no issues in 28 source files
- `uv run pytest -q` -- 2090 passed, 423 deselected
- Notebook executed through `jupyter nbconvert --execute` on the prepared subset: no errors, five figures, about three seconds wall time
- `scripts/prepare_fire_demo_inputs.py` ran end to end anonymously; cold cache downloads roughly a gigabyte of compressed chunks in about twelve minutes, a warm cache regenerates the published files in a few seconds

Fixes #811

## Summary by Sourcery

Provide a reproducible ERA5-based CONUS fire-weather demonstration with cached inputs, documented limitations, and comprehensive preparation tests.

New Features:
- Add an end-to-end ERA5 CONUS fire-weather demonstration notebook covering KBDI, CFFWIS, Fosberg FFWI, and HDW.
- Add a reproducible input-preparation workflow that downloads and caches anonymous ARCO-ERA5 subsets for the demonstration.

Enhancements:
- Document the demonstration’s scope, approximations, and distinction from operational reference products.

Build:
- Add aiohttp to the development dependencies for the ERA5 data preparation workflow.

Documentation:
- Link the fire-weather demonstration from the wildfire applications documentation and describe its intended use and limitations.

Tests:
- Add regression coverage for ERA5 aggregation, derived fields, caching, publishing, provenance manifests, and coordinate handling.

## Summary by CodeRabbit

- **New Features**
  - Added a CONUS fire-weather demonstration notebook covering KBDI, CFFWIS, Fosberg FFWI, and HDW calculations.
  - Added a workflow for preparing and caching daily ERA5 demonstration inputs, including derived humidity and height data.
  - Added visualizations for seasonal conditions, percentiles, timing, and a documented 7 September 2020 case study.

- **Documentation**
  - Added guidance on running the demonstration, its approximations, data sources, and chunking considerations.

- **Tests**
  - Added coverage for input preparation, aggregation, derived fields, caching, provenance, and failure handling.

### CODE REVIEW

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** https://github.com/monocongo/climate_indices/pull/981 (head `a72e09b2`, base `main` at `88a78fd8`)
- **Lanes run:** 6 sub-agent lanes total — 4 review lanes (correctness, dataflow, tests, maintainability) and 2 verification lanes re-run on the fix commits. All lanes ran on the deepseek-backed lane agents: the codex-backed lanes failed with usage-limit and model-availability errors and were retried on the available provider.
- **Validation:** `uv run ruff check src/ tests/ scripts/prepare_fire_demo_inputs.py` pass; `uv run ruff format --check src/ tests/ scripts/prepare_fire_demo_inputs.py` pass (93 files formatted); `uv run mypy src/` pass (28 source files); `uv run pytest -q` pass (2090 passed, 423 deselected). `jupyter nbconvert --execute` on the notebook: 0 errors, 5 figures, ~3 s. `scripts/prepare_fire_demo_inputs.py` ran end to end anonymously (cold cache ~12 min, warm cache seconds).
- **Fix commits:**
  - [2e9b4811 — fix(fire): correct demo input binning, provenance, and claims](https://github.com/monocongo/climate_indices/commit/2e9b4811)
  - [be4c8f61 — fix(fire): complete the year-end precipitation bin and the HDW caveat](https://github.com/monocongo/climate_indices/commit/be4c8f61)
  - [a72e09b2 — docs(fire): fix the demo notebook typo and the slice docstring wording](https://github.com/monocongo/climate_indices/commit/a72e09b2)

### Findings and fix status
#### [P2] KBDI map labelled in hundredths of an inch while the call returns metric millimetres — Fixed
- **Location:** `notebooks/fire_weather_demo.ipynb`, KBDI map cell
- **Origin:** dataflow
- **Problem:** `fire.kbdi()` was called without `units=`, so the adapter returned millimetres (`attrs["units"] == "mm"`, scale 0-203.2) while the colorbar said "hundredths of an inch".
- **Applied fix:** colorbar labelled `KBDI (mm, 0-203.2)`, commit `2e9b4811`.

#### [P2] Cold-started CFFWIS presented as multi-year drought memory — Fixed
- **Location:** `notebooks/fire_weather_demo.ipynb`, DC panel title and "What this demo does not claim"
- **Origin:** dataflow
- **Problem:** FFMC/DMC/DC start from the literature seeds (85/6/15) on 1 March 2020, but the panel was titled "Deep drought memory: DC" and the approximation section did not say so.
- **Applied fix:** panel retitled "DC at the end of the season"; a cold-start bullet added to the limitations section, commit `2e9b4811`.

#### [P3] Cache could silently reuse a download made for different settings — Fixed
- **Location:** `scripts/prepare_fire_demo_inputs.py`, `_cache`
- **Origin:** correctness
- **Problem:** the cache key held only the variable and year/season, so a changed source URL, domain, or window reused stale data with no check; the repository's `prepare_e2e_inputs.py` checksums downloads for exactly this reason.
- **Applied fix:** `_cache_fingerprint()` (hash of source URL, domain, years, season) is folded into every cache file name, and the per-year slice bounds are now spelled out in the key, commits `2e9b4811` and `be4c8f61`.

#### [P3] Pressure-level wind inherited the surface long name — Fixed
- **Location:** `scripts/prepare_fire_demo_inputs.py`, `_add_units`
- **Origin:** correctness
- **Problem:** `levels_daily_2020_season.nc` described its 13-level `wind_speed_ms` as "10 m wind speed".
- **Applied fix:** level-aware long name ("Wind speed at the pressure level"), commit `2e9b4811`.

#### [P3] Manifest recorded the requested window as if it were the realized grid — Fixed
- **Location:** `scripts/prepare_fire_demo_inputs.py`, manifest
- **Origin:** correctness
- **Problem:** the manifest listed the requested box (25-50 N, 235-295 E) without noting that the 1.5 degree grid realizes 25.5-49.5 N, 235.5-294.0 E, and the module docstring had a longitude typo.
- **Applied fix:** manifest now records `requested_*` and `realized_*` latitude/longitude computed from the published datasets; docstring corrected, commit `2e9b4811`.

#### [P3] 31 December precipitation bins were one 6-hour accumulation short — Fixed
- **Location:** `scripts/prepare_fire_demo_inputs.py`, `_surface_inputs` and `_to_daily_surface`
- **Origin:** correctness
- **Problem:** with the new accumulation-day binning, each year's six-hourly slice ended at 31 December 18:00 UTC, so the 31 December bin missed the accumulation stamped 1 January 00:00 and the annual totals were short at the boundary. (The lane's stronger symptoms — duplicate 31 December rows, NaN temperatures — did not reproduce in the published artifact; the row count, uniqueness, and NaNs were already clean.)
- **Applied fix:** each year's slice now carries the first day of the next year and the daily rows are trimmed back to the requested year; 31 December domain precipitation rose from 1852.1/1262.5/2256.2 mm to 2676.5/1656.8/2940.8 mm for 2018/2019/2020 while interior days are byte-identical; a year-boundary test was added, commit `be4c8f61`.

#### [P3] Notebook prose contradicted the execution model, the HDW logging, and the percentile promise — Fixed
- **Location:** `notebooks/fire_weather_demo.ipynb`
- **Origin:** dataflow
- **Problem:** the timing section called the arrays "eager NetCDF-backed" though the notebook opens them with `chunks={}`; the HDW paragraph promised a warning the Dask path suppresses; the introduction promised 90th/95th percentile maps but only the 95th was computed and `percentile_levels` was unused.
- **Applied fix:** prose corrected and the Fosberg cell now plots both 90th and 95th percentiles using `percentile_levels`, commits `2e9b4811` and `be4c8f61`.

#### [P3] Test gaps: cache, publish, manifest, coordinate conversion, and assertions that could not fail — Fixed
- **Location:** `tests/test_prepare_fire_demo_inputs.py`
- **Origin:** tests
- **Problem:** `_cache`, `_publish`, `_add_units`, `_select`, and the manifest were untested; the relative-humidity test sampled only clip boundaries (a broken kPa-to-hPa factor would still read 100 percent), the precipitation test never exercised the negative-noise clip, and the AGL test used a zero surface geopotential so deleting the subtraction would not fail it.
- **Applied fix:** tests added for cache hit/miss and failure cleanup, publish failure preserving the previous output, manifest provenance and checksums, longitude conversion and time-less fields, an interior 50 percent humidity point, a clipped negative accumulation, a non-zero surface geopotential, the static field's time axis, and the year-boundary bin; commit `2e9b4811` and `be4c8f61`.

#### [P3] Redundant `requests` dev dependency — Fixed
- **Location:** `pyproject.toml`
- **Origin:** maintainability
- **Problem:** neither the script nor the notebook imports `requests`; fsspec's HTTP backend imports `aiohttp`, so the second pin added nothing and `requests` was already present transitively.
- **Applied fix:** `requests` removed from the dev group, `aiohttp` retained, `uv.lock` consistent, commit `2e9b4811`.

#### [P4] Surface-relative-humidity assertion was bounds-only — Fixed
- **Location:** `tests/test_prepare_fire_demo_inputs.py`
- **Origin:** tests
- **Problem:** the integration test asserted only `0 <= x <= 100`, which passes for any wrong value in range.
- **Applied fix:** asserts the derived value (~91.2 percent) and equality with the selected level-0 humidity, commit `2e9b4811`.

#### [P4] Typos — Fixed
- **Location:** `notebooks/fire_weather_demo.ipynb` (GFWED name), `scripts/prepare_fire_demo_inputs.py` docstring
- **Origin:** dataflow, correctness
- **Problem:** "Global Fire WEather Database"; the docstring described the year-end slice as "one extra six-hourly stamp" when the slice actually carries the whole next day.
- **Applied fix:** both corrected, commit `a72e09b2`.

#### [P3] New script is outside CI's lint scope — Not fixed (precedent, not a defect)
- **Location:** `.github/workflows/unit-tests-workflow.yml`
- **Origin:** maintainability
- **Problem:** CI lints `src/` and `tests/` only, so the new script is covered by tests but not by CI lint.
- **Applied fix:** none; `scripts/prepare_e2e_inputs.py` and `scripts/generate_llms_txt.py` have the same status, so extending CI lint scope is a separate maintainer decision. Lint and format were run locally on the new script.

### Remaining blockers
None.

File: notebooks/fire_weather_demo.ipynb [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+    "\n",
+    "data_dir = Path(os.environ.get(\"CLIMATE_INDICES_FIRE_DEMO_DATA\", \"../data/fire-demo\"))\n",
+    "surface_path = data_dir / \"surface_daily_2018_2020.nc\"\n",
+    "levels_path = data_dir / \"levels_daily_2020_season.nc\"\n",
+    "for path in (surface_path, levels_path):\n",
+    "    if not path.exists():\n",
+    "        raise FileNotFoundError(\n",
+    "            f\"Prepared input not found: {path}. \"\n",
+    "            \"Generate it with: uv run --group dev scripts/prepare_fire_demo_inputs.py\"\n",
+    "        )\n",
```
[the remark this answers is not in the document]
Comment (monocongo) [reply notebooks/fire_weather_demo.ipynb]: Addressed in 1164328a: the guard now checks `data_dir / "manifest.json"` alongside the NetCDF inputs, so a missing manifest raises the same guided `FileNotFoundError` before cell-04 reads it.

File: notebooks/fire_weather_demo.ipynb [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+  {
+   "cell_type": "code",
+   "id": "cell-16",
+   "metadata": {},
+   "source": [
+    "fig, axes = plt.subplots(1, 3, figsize=(16, 4))\n",
+    "cffwis[\"dc\"].isel(time=-1).plot(ax=axes[0], cmap=\"YlOrBr\", cbar_kwargs={\"label\": \"DC\"})\n",
+    "axes[0].set_title(\"DC at the end of the season\")\n",
+    "kbdi.quantile(0.95, dim=\"time\").plot(ax=axes[1], cmap=\"YlOrRd\", cbar_kwargs={\"label\": \"KBDI\"})\n",
+    "axes[1].set_title(\"95th percentile KBDI\")\n",
```
[the remark this answers is not in the document]
Comment (monocongo) [reply notebooks/fire_weather_demo.ipynb]: Addressed in 1164328a: `kbdi` is filtered with `.sel(time=slice(season_start, season_end))` before `quantile(0.95, dim="time")`, so the KBDI panel ranks the same 2020 season as the neighboring FWI panels.

File: scripts/prepare_fire_demo_inputs.py [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+        default=Path("data/fire-demo"),
+        help="directory that receives the prepared NetCDF files and manifest (default: data/fire-demo)",
+    )
+    arguments = parser.parse_args()
+    output_dir: Path = arguments.output_dir
+    manifest = prepare_inputs(output_dir)
+    manifest_path = output_dir / "manifest.json"
+    temporary = manifest_path.with_suffix(f".json.{uuid.uuid4().hex}")
+    temporary.write_text(json.dumps(manifest, indent=2) + "\n")
+    shutil.move(str(temporary), manifest_path)
```
[the remark this answers is not in the document]
Comment (monocongo) [reply scripts/prepare_fire_demo_inputs.py]: Addressed in 1164328a: the manifest is published with `os.replace` inside the same `try/finally` cleanup as `_publish`, and the now-unused `shutil` import is removed.

File: scripts/prepare_fire_demo_inputs.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+def _open_source() -> xr.Dataset:
+    """Open the ARCO-ERA5 store lazily over anonymous HTTPS."""
+    return xr.open_zarr(SOURCE_URL, chunks={}, consolidated=True, decode_times=True)
+
+
+def _select(dataset: xr.Dataset, name: str, start: str, end: str) -> xr.DataArray:
+    """Subset one variable to the CONUS box and inclusive time window."""
+    variable = dataset[name]
+    selection: dict[str, slice] = {
+        "latitude": slice(*DOMAIN_LATITUDE),
```
[the remark this answers is not in the document]
Comment (monocongo) [reply scripts/prepare_fire_demo_inputs.py:110]: Checked this against the live store: `latitude` on `1959-2022-6h-240x121_equiangular_with_poles_conservative.zarr` is monotonically **increasing**, -90 to 90 (not descending 90 to -90 as stated).

```
>>> ds['latitude'].values[:5]
array([-90. , -88.5, -87. , -85.5, -84. ])
>>> ds['latitude'].values[-5:]
array([84. , 85.5, 87. , 88.5, 90. ])
```

Running `_select` against the real store confirms `slice(25.0, 50.0)` on this ordering returns the expected 17 CONUS latitudes (25.5-49.5 N), not an empty selection:

```
>>> _select(ds, '2m_temperature', '2018-01-01', '2018-01-02').sizes
Frozen({'time': 8, 'longitude': 40, 'latitude': 17})
```

No code change made — the premise (descending coordinate) doesn't hold for this source, so the reported failure mode doesn't reproduce.

File: scripts/prepare_fire_demo_inputs.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    """Aggregate the six-hourly surface variables of one year to a daily dataset."""
+    daily = xr.Dataset()
+    temperature = dataset["2m_temperature"] - 273.15
+    daily["tmean_c"] = temperature.resample(time="1D").mean()
+    daily["tmax_c"] = temperature.resample(time="1D").max()
+    daily["tmin_c"] = temperature.resample(time="1D").min()
+    # ERA5 accumulates precipitation over the preceding six hours; summing the
+    # four daily accumulations gives the daily total. Tiny negative values are
+    # numerical noise and would fail the fire recurrences' non-negativity check.
+    daily["precip_mm"] = (dataset["total_precipitation_6hr"] * 1000.0).clip(min=0.0).resample(time="1D").sum()
```
[the remark this answers is not in the document]
Comment (monocongo) [reply scripts/prepare_fire_demo_inputs.py]: Fixed on this branch, in commits `2e9b4811` and `be4c8f61`:

- `2e9b4811` bins the 6-hour accumulations with `resample(time="1D", closed="right", label="left")` so the four accumulations ending 06/12/18/24 UTC land on the calendar day they cover, and documents the convention in the manifest (`"sum of the four 6-hour accumulations ending 06, 12, 18, and 24 UTC"`).
- `be4c8f61` closes the resulting gap at each year boundary: the per-year slice now extends one stamp into the next year so 31 December's bin is complete, then trims the daily rows back to the requested year.

File: scripts/prepare_fire_demo_inputs.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+        _add_units(dataset).to_netcdf(temporary)
+        os.replace(temporary, path)
+    finally:
+        temporary.unlink(missing_ok=True)
+
+
+def prepare_inputs(output_dir: Path) -> dict[str, Any]:
+    """Prepare and publish the demonstration inputs, returning the manifest."""
+    cache_dir = output_dir / "cache"
+    source = _open_source()
```
[the remark this answers is not in the document]
Comment (monocongo) [reply scripts/prepare_fire_demo_inputs.py]: Fixed in `0812cbb0`: `_open_source()` is now called lazily through a closure passed into `_surface_inputs`/`_level_inputs`, which only invoke it from a `_cache` builder on an actual cache miss. A run where every selection is already cached never opens the remote store. Covered by `test_prepare_inputs_skips_opening_source_when_fully_cached`, which fails `_open_source` if it's called and asserts `prepare_inputs` still completes.

File: scripts/prepare_fire_demo_inputs.py [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+    try:
+        surface = _surface_inputs(source, cache_dir)
+        levels = _level_inputs(source, cache_dir)
+    finally:
+        source.close()
+
+    surface_path = output_dir / OUTPUT_SURFACE
+    levels_path = output_dir / OUTPUT_LEVELS
+    _publish(surface, surface_path)
```
[the remark this answers is not in the document]
Comment (monocongo) [reply scripts/prepare_fire_demo_inputs.py:362]: Addressed in `0812cbb0` on the consumer side, per the finding's own framing ("either a mutually consistent completed artifact set or a way to reject a mismatch before computation"): the manifest already records each artifact's sha256, so the notebook now recomputes and compares both hashes right after loading the manifest, before any index is computed, and raises with a regeneration hint on a mismatch. That's the earliest point in the notebook the check can run, since `surface_path`/`levels_path` are opened lazily in the cell above and nothing is materialized until later cells.

Went with hash verification rather than restructuring `prepare_inputs` to publish the two NetCDF files and manifest as a single atomic swap, since the latter is a larger change to the publish pipeline for the same guarantee the manifest already gives the consumer for free.
